src/main/java/com/odysseusinc/arachne/datanode/controller/BaseDataSourceController.java

```diff
@@ -121,20 +122,56 @@ protected BaseDataSourceController(UserService userService,
                                          @Valid @RequestPart("dataSource") CreateDataSourceDTO dataSourceDTO,
                                          @RequestPart(name = "keyfile", required = false) MultipartFile keyfile,
                                          BindingResult bindingResult
-    ) throws NotExistException, PermissionDeniedException {
+    ) throws NotExistException {
+
+        return validatePersistDataSourceRequest(bindingResult, dataSourceDTO.getName(), null)
+                .orElseGet(() -> saveDataSource(null, dataSourceDTO, principal, keyfile, dataSourceService::create));
+    }
+
+    @ApiOperation(value = "Updates given data source")
+    @RequestMapping(
+            value = Constants.Api.DataSource.UPDATE,
+            method = RequestMethod.PUT
+    )
+    public JsonResult<DataSourceDTO> update(Principal principal,
+                                            @Valid @RequestPart("dataSource") CreateDataSourceDTO dataSourceDTO,
+                                            @PathVariable("id") Long id,
+                                            @RequestPart(name = "keyfile", required = false) MultipartFile keyfile,
+                                            BindingResult bindingResult)
+            throws PermissionDeniedException {
+
+        return validatePersistDataSourceRequest(bindingResult, dataSourceDTO.getName(), id)
+                .orElseGet(() -> saveDataSource(id, dataSourceDTO, principal, keyfile, dataSourceService::update));
+    }
+
+    private JsonResult saveDataSource(Long dataSourceId, CreateDataSourceDTO dataSourceDTO, Principal principal, MultipartFile keyfile, BiFunction<User, DataSource, DataSource> persistDatasource) {
 
-        if (bindingResult.hasErrors()) {
-            return setValidationErrors(bindingResult);
-        }
         final User user = getAdmin(principal);
         dataSourceDTO.setKeyfile(keyfile);
         DataSource dataSource = conversionService.convert(dataSourceDTO, DataSource.class);
-        DataSource optional = dataSourceService.create(user, dataSource);
+        dataSource.setId(dataSourceId);
+
+        DataSource persistedDataSource = persistDatasource.apply(user, dataSource);
         JsonResult<DataSourceDTO> result = new JsonResult<>(NO_ERROR);
-        result.setResult(conversionService.convert(optional, DataSourceDTO.class));
+        result.setResult(conversionService.convert(persistedDataSource, DataSourceDTO.class));
         return result;
     }
 
+    private Optional<JsonResult> validatePersistDataSourceRequest(BindingResult bindingResult, String dataSourceName, Long dataSourceId){
```

Comment:
Could you parameterize method to avoid warnings about unchecked assignments?

Reply:
@acumarav could you make a parameterized method?

Reply:
done